go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+			*nullCount = 0
+		} else {
+			var (
+				toWrite       int64
+				spacedToWrite int64
+			)
+			for i := int64(0); i < batchSize; i++ {
+				if defLevels[i] == w.levelInfo.DefLevel {
+					toWrite++
+				}
+				if defLevels[i] >= w.levelInfo.RepeatedAncestorDefLevel {
+					spacedToWrite++
+				}
+			}
+			*valuesToWrite += toWrite
+			*spacedValuesToWrite += spacedToWrite
+			*nullCount = spacedToWrite - toWrite
+		}
+		return
+	}
+
+	// shrink to fit possible causes another allocation
+	newBitmapSize := bitutil.BytesForBits(batchSize)
+	if newBitmapSize != int64(w.bitsBuffer.Len()) {
+		w.bitsBuffer.ResizeNoShrink(int(newBitmapSize))
+	}
+
+	io := ValidityBitmapInputOutput{
+		ValidBits:      w.bitsBuffer.Bytes(),
+		ReadUpperBound: batchSize,
+	}
+	DefLevelsToBitmap(defLevels[:batchSize], w.levelInfo, &io)
+	*valuesToWrite = io.Read - io.NullCount
+	*spacedValuesToWrite = io.Read
+	*nullCount = io.NullCount
+}
+
+func (w *columnWriter) getPageStatistics() (enc metadata.EncodedStatistics, err error) {
+	if w.pageStatistics != nil {
+		enc, err = w.pageStatistics.Encode()
+	}
+	return
+}
+
+func (w *columnWriter) getChunkStatistics() (enc metadata.EncodedStatistics, err error) {
+	if w.chunkStatistics != nil {
+		enc, err = w.chunkStatistics.Encode()
+	}
+	return
+}
+
+func (w *columnWriter) resetPageStatistics() {
+	if w.chunkStatistics != nil {
+		w.chunkStatistics.Merge(w.pageStatistics)
+		w.pageStatistics.Reset()
+	}
+}
+
+func (w *columnWriter) Close() (err error) {
+	if !w.closed {
+		w.closed = true
+		if w.hasDict && !w.fallback {
+			w.WriteDictionaryPage()
+		}
+
+		w.FlushBufferedDataPages()
+
+		var chunkStats metadata.EncodedStatistics
+		chunkStats, err = w.getChunkStatistics()
+		if err != nil {
+			return err
+		}
+
+		chunkStats.ApplyStatSizeLimits(int(w.props.MaxStatsSizeFor(w.descr.Path())))
+		chunkStats.Signed = schema.SortSIGNED == w.descr.SortOrder()
+
+		if w.rowsWritten > 0 && chunkStats.IsSet() {
+			w.metaData.SetStats(chunkStats)
+		}
+		err = w.pager.Close(w.hasDict, w.fallback)
+
+		w.defLevelSink.Reset(0)
+		w.repLevelSink.Reset(0)
+	}
+	return err
+}
+
+func doBatches(total, batchSize int64, action func(offset, batch int64)) {
+	numBatches := total / batchSize
+	for i := int64(0); i < numBatches; i++ {
+		action(i*batchSize, batchSize)
+	}
+	if total%batchSize > 0 {
+		action(numBatches*batchSize, total%batchSize)
+	}
+}
+
+func levelSliceOrNil(rep []int16, offset, batch int64) []int16 {
+	if rep == nil {
+		return nil
+	}
+	return rep[offset : batch+offset]
+}
+
+func (w *ByteArrayColumnChunkWriter) maybeReplaceValidity(values array.Interface, newNullCount int64) array.Interface {
+	if w.bitsBuffer == nil {
+		return values
+	}
+
+	buffers := values.Data().Buffers()
+	if len(buffers) == 0 {
+		return values
+	}
+
+	if values.Data().Offset() > 0 {
+		data := values.Data()
+		buffers[1] = memory.NewBufferBytes(data.Buffers()[1].Bytes()[data.Offset()*arrow.Int32SizeBytes : data.Len()*arrow.Int32SizeBytes])
```

Comment:
this might be based on old buggy code, please take a look at: https://github.com/apache/arrow/blob/040bd1e8d563bed532b712721078757f4aa0907e/cpp/src/parquet/column_writer.cc#L1321

I think the actual validity buffer needs to be sliced as well.

Reply:
looking at the file you linked, it looks like what i was missing was that i needed to do `buffers[0] = w.bitsBuffer` which is what the code linked does. Which makes sense as the `w.bitsBuffer` should already be updated with the correct slice of the validity buffer we want.

Reply:
yeah, please look at the blame and look at the additional unit tests added to make sure it is covered.